### ISSUE TITLE
Fix runc issues to enable container reuse

### ISSUE THREAD

**Issue by dgrove-oss** [apache/incubator-openwhisk-deploy-kube#67]
I'm seeing similar errors as reported in https://github.com/apache/incubator-openwhisk/issues/2788 with docker-runc getting a Go deserialization error attempting to read the state.json for a container when executing in kubernetes.
```
[2017-09-29T20:15:54.551Z] [ERROR] [#sid_102] [RuncClient] code: 1, stdout: , stderr: json: cannot unmarshal object into Go value of type []string [marker:invoker_runc.pause_error:6830148:259]
```

I think the right fix for kubernetes might be to setup a sidecar container with the appropriate docker version so we keep all of the action containers contained within the invoker pod.   See https://applatix.com/case-docker-docker-kubernetes-part-2/ for some discussion.  This would get better confinement of the containers to the pod.   An alternative fix would be to figure out how to mount the runc from the workernode.

Attaching a shell to the invoker pod, I can reproduce the docker-runc pause error manually.  I can also see that docker pause/unpause work as expected if I use those commands.

In any case, we are currently not getting container reuse with kubernetes via runc currently with the default InvokerReactive so performance is not what we would expect.

**Comment by csantanapr** [comment]
@ddragosd is being looking into this trying multiple combinations to mount docker and docker-runc 

Related to https://github.com/apache/incubator-openwhisk/pull/2799

**Comment by ddragosd** [comment]
> An alternative fix would be to figure out how to mount the runc from the workernode.

This is what apache/incubator-openwhisk#2799 does. Initially we tried mounting both `docker` and `runc`, and ended up only mounting `runc` b/c `docker` binary may have other dependencies on the host as well, but the way `runc` is built today didn't show any errors in that regard. I'm not sure how many guarantees we have in the future that this premise won't break, but so far the tests are successful.

**Comment by dgrove-oss** [comment]
Right.  The think I'm wondering about is if we want to contain all of the action containers within the invoker pod or if we want to allow them to leak out to the wortkernode.  If we want them contained, then we run a docker demon in a sidecar within the pod.  If we want to let them flow out to the workernode, then we attempt to mount the runc binary from the workernode within the invoker pod.

**Comment by ddragosd** [comment]
> a docker demon in a sidecar within the pod

how will this sidecar daemon communicate with the host ?

**Comment by dgrove-oss** [comment]
It's the situation of Case 2: Pods and DinD in the blog post I referenced above.  As I understand it, the sidecar doesn't communicate with the host docker at all.  It's running its own contained world of containers within the pod that is being run by the host docker.   It at least seems like an interesting architecture to play with, so I will likely try to prototype it next week.

**Comment by ddragosd** [comment]
This is an interesting idea indeed. I'm curious to see the outcome of your POC, when you have something working, in particular any perf improvements (if any) on container startup, pause, stop, rm. 

With the sidecard there is still a dependency between the invoker container and its sidecar... which makes me wonder, why not make the 2 one and the same: have invoker act and its own sidecar, with its own docker version, w/o mounting anything from the host at all, not even `docker.sock`

**Comment by dgrove-oss** [comment]
Merging the two is an interesting idea.  It does require the invoker image to have docker installed and running within it (not unreasonable, but different than current practice), but in return you get a completely contained and reliable environment.  Probably worth trying it out both with separate sidecar container and docker merged in and see what it looks like and how it performs.

**Comment by markusthoemmes** [comment]
FWIW: When I last checked docker-in-docker was possible, but not stable nor reliable at all. It has been thought of but not sure if there ever was a PoC. Might be worth a try.

**Comment by dgrove-oss** [comment]
yeah, I did some more reading about docker-in-docker over the weekend and I'm less excited about that option now.  It also exposes the invoker pod-internal network to the action container and that's probably not what we actually want.

**Comment by dgrove-oss** [comment]
Poked a little at mounting docker-runc from the kube worker node this afternoon.  Figured out one problem I was having earlier this week is that my worker node is ubuntu 16.04 (not 14.04).  So libapparmor.so.1 moved to /lib/....   Fixed the hostpath in invoker.yaml and it mounted correctly.  Will have to do the same thing for a couple more libraries, but maybe it is viable.

**Comment by dgrove-oss** [comment]
I got docker-runc mounted from the kube worker node to execute and got past the Go deserialization errors.  However, it still fails to actually pause/resume the containers successfully.  Pushed the changes to invoker.yaml to https://github.com/dgrove-oss/incubator-openwhisk-deploy-kube/tree/mount-docker-runc to document them.

I also tried moving the invoker image to be based on 16.04 instead of 14.04.  That did not help either (although it allowed the elimination of mounting the libapparmor and libseccomp libraries). 

For now, I'm putting trying to get docker-runc to work on hold.  Once https://github.com/apache/incubator-openwhisk/pull/2828 gets merged, I'll change invoker.yaml to set INVOKER_USE_RUNC to false and we can use docker to pause/unpause until it is clear that is really a performance problem in the kube environment.
